pages/howto/containerization.md

```diff
@@ -0,0 +1,163 @@
+```
+
+Then run, execute, or shell into the container as normal.
+
+### Building a SIF image
+
+In the event that a container is unavailable for a given application, you may need to build your own container from scratch. You will need to download and install Apptainer on your own machine. Instructions to do this can be found in the [documentation for Apptainer](https://apptainer.org/docs/admin/main/installation.html).
+
+Apptainer allows a user to build images using a *definition file*. Just like a Dockerfile, this file has a variety of directives that allow for the customization of your image. A sample image would look something like this: 
+
+```
+Bootstrap: docker
+From: ubuntu:20.04
+
+%help
+	I am help text!
+
+%setup		
+	apt-get update
+	apt-get install nano
+	apt-get install gcc 
+
+%runscript
+	echo “hello! I am a container!”
+```
+
+#### Apptainer Build
+
+Once you have written your Apptainer definition file, you can build the application locally with the `apptainer build` command, as follows:
+
+```
+apptainer build <localname>.sif <recipe-name>.def
+```
+
+### GPU-enabled Containers 
+
+It is possible to run GPU workloads within singularity containers. This will allow you to run more up-to-date versions of software like PyTorch which may have older versions installed as modules. To do so, merely add the `--nv` flag when you use `run` or `exec` commands like so:
+
+`singularity run --nv <image_name>.sif`
+
+NVIDIA also hosts a [number of containers]() as part of their own container library. These can be pulled and run by singularity, though the pulling process can take several hours depending on the image size.
+
+### Building MPI-enabled images
+MPI-enabled Apptainer containers can be deployed on CCR's systems with the caveat that the MPI software within the container has a similar (not necessarily exact) version with MPI software available on the system. This requirement diminishes the portability of MPI-enabled containers, as they may not run on other systems without compatible MPI software. Regardless, MPI-enabled containers can still be a very useful option in many cases. 
+
+Here we provide an example of using a gcc compiler with OpenMPI. CCR's system uses an Infiniband interconnect. In order to use a Singularity container with OpenMPI (or any MPI) on the cluster, OpenMPI needs to be installed both inside and outside of the container. More specifically, the _same_ version of OpenMPI needs to be installed inside and outside (at least very similar, you can sometimes get away with two different minor versions, e.g. 2.1 and 2.0). 
+
+
+Once you’ve built the container with one of the methods outlined above, you can place it in your home directory and run it on a compute node. The following is an example of running a gcc/OpenMPI container with Apptainer. The syntax is a normal MPI run where multiple instances of a Singularity image are run. The following example runs `mpi_hello_world` with MPI from a container.
```

Comment:
```suggestion
Once you’ve built the container with one of the methods outlined above, you can place it in your home or project directory and run it on a compute node. The following is an example of running a gcc/OpenMPI container with Apptainer. The syntax is a normal MPI run where multiple instances of a Singularity image are run. The following example runs `mpi_hello_world` with MPI from a container.
```

Reply:
I like the change, though I will say I'm not sure 'mpi_hello_world' exists as a container. I figured the example syntax was still helpful, though it may help to have a hello world container down the line.